File: server/bot/src/integration/test/clan_test.cpp
#include <fb/bot/integration/clan_test.h>
#include <fb/bot/integration/game_controller.h>
#include <fb/logger.h>
#include <format>

using namespace std::chrono_literals;
using namespace fb::bot::integration;

namespace game_reqs = fb::protocol::game::request;
namespace game_resp = fb::protocol::game::response;

clan_test::clan_test(game_bot_controller& controller) :
    bot_integration_test(controller, 4)
{ }

async::task<void> clan_test::on_initialize(game_bot_controller& controller)
{
    co_await super::on_initialize(controller);
    co_await super::arrange_bots_in_line_formation();
    co_return;
}

async::task<void> clan_test::on_scenario_started(uint32_t scenario_index)
{
    co_return;
}

async::task<void> clan_test::on_scenario_finished(uint32_t scenario_index)
{
    fb::logger::debug("Clan test scenario {} finished", scenario_index);
    co_return;
}

async::task<bool> clan_test::run_script_step(const std::shared_ptr<game_bot>&        bot,
                                             std::string_view                        step,
                                             std::initializer_list<std::string_view> extra_args)
{
    auto cmd         = std::format("/스크립트 {} run {}", SCRIPT_FILE, step);
    auto step_str    = std::string(step);
    auto pass_prefix = std::format("CL:PASS:{}", step_str);
    auto fail_prefix = std::format("CL:FAIL:{}", step_str);

    for (auto arg : extra_args)
        cmd.append(std::format(" {}", arg));

    try
    {
        auto&& resp = co_await bot->request<game_resp::chat>(
            game_reqs::chat(false, cmd),
            [pass_prefix, fail_prefix](auto& response) -> bool {
                return response.text.find(pass_prefix) != std::string::npos ||
                       response.text.find(fail_prefix) != std::string::npos;
            },
            DEFAULT_TIMEOUT);

        if (resp.text.find(pass_prefix) == std::string::npos)
        {
            fb::logger::fatal("Clan script step failed ({}): {}", step_str, resp.text);
            co_return false;
        }

        co_return true;
    }
    catch (const std::exception& e)
    {
        fb::logger::fatal("Clan script step error ({}): {}", step_str, e.what());
        co_return false;
    }
}

async::task<bool> clan_test::test_clan_creation()
{
    auto bots = this->get_test_bots();
    bots[0]->chat("=== CLAN CREATION TEST STARTED ===");

    if (!co_await this->run_script_step(bots[0], "create_clan"))
        co_return false;

    bots[0]->chat("=== CLAN CREATION TEST COMPLETED SUCCESSFULLY ===");
    co_return true;
}

async::task<bool> clan_test::test_clan_title()
{
    auto bots = this->get_test_bots();
    bots[0]->chat("=== CLAN TITLE TEST STARTED ===");

    if (!co_await this->run_script_step(bots[0], "set_title"))
        co_return false;

    bots[0]->chat("=== CLAN TITLE TEST COMPLETED SUCCESSFULLY ===");
    co_return true;
}

async::task<bool> clan_test::test_clan_invite()
{
    auto bots = this->get_test_bots();
    bots[0]->chat("=== CLAN INVITE TEST STARTED ===");

    if (!co_await this->run_script_step(bots[0], "invite", {bots[1]->name()}))
        co_return false;

    if (!co_await this->run_script_step(bots[1], "verify_member", {bots[0]->name()}))
        co_return false;

    bots[0]->chat("=== CLAN INVITE TEST COMPLETED SUCCESSFULLY ===");
    co_return true;
}

async::task<bool> clan_test::test_clan_role()
{
    auto bots = this->get_test_bots();
    bots[0]->chat("=== CLAN ROLE TEST STARTED ===");

    if (!co_await this->run_script_step(bots[1], "invite_fail", {bots[2]->name()}))
        co_return false;

    if (!co_await this->run_script_step(bots[0], "promote_deputy", {bots[1]->name()}))
        co_return false;

    if (!co_await this->run_script_step(bots[1], "invite", {bots[2]->name()}))
        co_return false;

    if (!co_await this->run_script_step(bots[1], "change_role_fail", {bots[2]->name(), "3"}))
        co_return false;

    if (!co_await this->run_script_step(bots[1], "change_role_fail", {bots[3]->name(), "0"}))
        co_return false;

    bots[0]->chat("=== CLAN ROLE TEST COMPLETED SUCCESSFULLY ===");
    co_return true;
}

async::task<bool> clan_test::test_clan_title_change()
{
    auto bots = this->get_test_bots();
    bots[0]->chat("=== CLAN TITLE CHANGE TEST STARTED ===");

    if (!co_await this->run_script_step(bots[2], "change_title_fail", {"MateTitle"}))
        co_return false;

    if (!co_await this->run_script_step(bots[1], "change_title_fail", {"DeputyTitle"}))
        co_return false;

    if (!co_await this->run_script_step(bots[0], "change_title", {"MasterTitle"}))
        co_return false;

    bots[0]->chat("=== CLAN TITLE CHANGE TEST COMPLETED SUCCESSFULLY ===");
    co_return true;
}

async::task<bool> clan_test::test_clan_disbanding()
{
    auto bots = this->get_test_bots();
    bots[0]->chat("=== CLAN DISBANDING TEST STARTED ===");

    if (!co_await this->run_script_step(bots[2], "kick_fail", {bots[1]->name()}))
        co_return false;

    if (!co_await this->run_script_step(bots[2], "leave"))
        co_return false;

    if (!co_await this->run_script_step(bots[0], "kick", {bots[1]->name()}))
        co_return false;

    if (!co_await this->run_script_step(bots[0], "destroy_clan"))
        co_return false;

    bots[0]->chat("=== CLAN DISBANDING TEST COMPLETED SUCCESSFULLY ===");
    co_return true;
}

fb::generator<bot_integration_test::scenario_t> clan_test::on_generate_scenario()
{
    co_yield [this]() -> async::task<bool> {
        co_return co_await this->test_clan_creation();
    };
    co_yield [this]() -> async::task<bool> {
        co_return co_await this->test_clan_title();
    };
    co_yield [this]() -> async::task<bool> {
        co_return co_await this->test_clan_invite();
    };
    co_yield [this]() -> async::task<bool> {
        co_return co_await this->test_clan_role();
    };
    co_yield [this]() -> async::task<bool> {
        co_return co_await this->test_clan_title_change();
    };
    co_yield [this]() -> async::task<bool> {
        co_return co_await this->test_clan_disbanding();
    };
}

std::string clan_test::name() const
{
    return "Clan Test";
}

File: server/game/scripts/clan_test.lua
local ROLE_DEPUTY = 2

local function cl_chat(me, step, ok, detail)
    if ok then
        me:chat('CL:PASS:' .. step, CHAT_TYPE.NORMAL, false)
    else
        me:chat('CL:FAIL:' .. step .. ':' .. (detail or ''), CHAT_TYPE.NORMAL, false)
    end
end

local function cl_assert(me, step, ok, detail)
    cl_chat(me, step, ok, detail)
    return ok
end

local function get_clan(me)
    local clan = me:clan()
    if clan == nil then
        return nil, 'no clan'
    end
    return clan
end

local function join_member(me, target_name)
    local clan, err = get_clan(me)
    if clan == nil then
        return err
    end
    return clan:join(me, target_name)
end

local function kick_member(me, target_name)
    local clan, err = get_clan(me)
    if clan == nil then
        return err
    end
    return clan:kick(me, target_name)
end

local function change_member_role(me, target_name, role)
    local clan, err = get_clan(me)
    if clan == nil then
        return err
    end
    return clan:change_role(me, target_name, role)
end

local function change_clan_title(me, title)
    local clan, err = get_clan(me)
    if clan == nil then
        return err
    end
    return clan:title(me, title)
end

function run(me, step, ...)
    local args = { ... }

    if step == 'create_clan' then
        if me:clan() ~= nil then
            return cl_assert(me, step, false, 'already in clan')
        end
        local err = me:create_clan(me:name())
        if err ~= nil then
            return cl_assert(me, step, false, err)
        end
        local clan = me:clan()
        if clan == nil or clan:name() ~= me:name() then
            return cl_assert(me, step, false, 'clan name mismatch')
        end
        return cl_assert(me, step, true)

    elseif step == 'set_title' then
        local title = me:name() .. '타이틀'
        local err = change_clan_title(me, title)
        if err ~= nil then
            return cl_assert(me, step, false, err)
        end
        local clan = me:clan()
        if clan == nil then
            return cl_assert(me, step, false, 'no clan')
        end
        if clan:title() ~= title then
            return cl_assert(me, step, false, 'title mismatch')
        end
        return cl_assert(me, step, true)

    elseif step == 'invite' then
        local target_name = args[1]
        if target_name == nil or target_name == '' then
            return cl_assert(me, step, false, 'missing target')
        end
        local err = join_member(me, target_name)
        if err ~= nil then
            return cl_assert(me, step, false, err)
        end
        return cl_assert(me, step, true)

    elseif step == 'invite_fail' then
        local target_name = args[1]
        if target_name == nil or target_name == '' then
            return cl_assert(me, step, false, 'missing target')
        end
        local err = join_member(me, target_name)
        if err == nil then
            return cl_assert(me, step, false, 'expected invite failure')
        end
        return cl_assert(me, step, true)

    elseif step == 'verify_member' then
        local expected_clan_name = args[1]
        if expected_clan_name == nil or expected_clan_name == '' then
            return cl_assert(me, step, false, 'missing clan name')
        end
        local clan = me:clan()
        if clan == nil then
            return cl_assert(me, step, false, 'not in clan')
        end
        if clan:name() ~= expected_clan_name then
            return cl_assert(me, step, false, 'clan name mismatch')
        end
        return cl_assert(me, step, true)

    elseif step == 'promote_deputy' then
        local target_name = args[1]
        if target_name == nil or target_name == '' then
            return cl_assert(me, step, false, 'missing target')
        end
        local err = change_member_role(me, target_name, ROLE_DEPUTY)
        if err ~= nil then
            return cl_assert(me, step, false, err)
        end
        return cl_assert(me, step, true)

    elseif step == 'change_role_fail' then
        local target_name = args[1]
        local role = tonumber(args[2])
        if target_name == nil or target_name == '' or role == nil then
            return cl_assert(me, step, false, 'missing arguments')
        end
        local err = change_member_role(me, target_name, role)
        if err == nil then
            return cl_assert(me, step, false, 'expected role change failure')
        end
        return cl_assert(me, step, true)

    elseif step == 'change_title' then
        local title = args[1]
        if title == nil or title == '' then
            return cl_assert(me, step, false, 'missing title')
        end
        local err = change_clan_title(me, title)
        if err ~= nil then
            return cl_assert(me, step, false, err)
        end
        local clan = me:clan()
        if clan == nil or clan:title() ~= title then
            return cl_assert(me, step, false, 'title mismatch')
        end
        return cl_assert(me, step, true)

    elseif step == 'change_title_fail' then
        local title = args[1]
        if title == nil or title == '' then
            return cl_assert(me, step, false, 'missing title')
        end
        local err = change_clan_title(me, title)
        if err == nil then
            return cl_assert(me, step, false, 'expected title change failure')
        end
        return cl_assert(me, step, true)

    elseif step == 'kick' then
        local target_name = args[1]
        if target_name == nil or target_name == '' then
            return cl_assert(me, step, false, 'missing target')
        end
        local err = kick_member(me, target_name)
        if err ~= nil then
            return cl_assert(me, step, false, err)
        end
        return cl_assert(me, step, true)

    elseif step == 'kick_fail' then
        local target_name = args[1]
        if target_name == nil or target_name == '' then
            return cl_assert(me, step, false, 'missing target')
        end
        local err = kick_member(me, target_name)
        if err == nil then
            return cl_assert(me, step, false, 'expected kick failure')
        end
        return cl_assert(me, step, true)

    elseif step == 'leave' then
        local clan, err = get_clan(me)
        if clan == nil then
            return cl_assert(me, step, false, err)
        end
        err = clan:leave(me)
        if err ~= nil then
            return cl_assert(me, step, false, err)
        end
        if me:clan() ~= nil then
            return cl_assert(me, step, false, 'still in clan')
        end
        return cl_assert(me, step, true)

    elseif step == 'destroy_clan' then
        local err = me:destroy_clan()
        if err ~= nil then
            return cl_assert(me, step, false, err)
        end
        if me:clan() ~= nil then
            return cl_assert(me, step, false, 'clan still exists')
        end
        return cl_assert(me, step, true)
    end

    return cl_assert(me, step, false, 'unknown step')
end
